### DISCUSSION TITLE
OpenTofu/Terraform modules not being pulled on run

### DISCUSSION THREAD

**Discussion by StoffelCPR** [semaphoreui/semaphore#2638] [Q&A] [answered]
Hey,

So we've been evaluating Semaphore a bit and want to see if it can be a good replacement for AWX.

One main advantage would be an automation platform for our OpenTofu code. Now in our Tofu code we have build our own TF module and put it into our git.

Now when I run an OpenTofu task it starts but fails during init phase because it can't pull the modules from our git.

Is this a bug or are we doing it wrong?

![image](https://github.com/user-attachments/assets/53c499d3-a9fa-41d7-b3da-f388d9eac411)

**Comment by fiftin** [comment, marked as answer]
Hi @StoffelCPR 

It is a really good question. I think it should be issue.

Do you have same SSH key for all modules?

If so and it is main use case we can add option to Template/Workspace with SSH key which will be used for cloning modules.

But if you have different keys for modules it will not work.

It is why Terraform Cloud and GitLab has its own modules registry. Looks like we need to implement it too. Is it suitable for you to import modules to Semaphore?

UPDATE:

Added support of SSH key for cloning modules in `v2.11.0-rc1`.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/0a6a0b4d-8b10-41df-8500-e3084d5b6c64" />

**Reply by StoffelCPR** [reply]
Yes, same key for all repos. We'll try the release candidate. Opened an issue shortly after creating this
